FOCUS #181: Usage Unit Dimension (#184)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/attributes/unit_format.md

# Unit Format

> Cost and utilization data frequently captures data measured in units related to data size, count, time, and other dimensions. The Unit Format attribute provides a standard for expressing units of measure in columns appearing in billing data.

Reviewer: ```suggestion
Cost and usage data frequently captures data measured in units related to data size, count, time, and other dimensions. The Unit Format attribute provides a standard for expressing units of measure in columns appearing in billing data.
```

> All columns defined in FOCUS specifying Unit Format as a value format MUST follow the requirements listed below.

Reviewer 2: I'm totally fine with this caveat, but if you prefer to not have it, we could also just call out PricingMeasurementUnit as an exception to this rule with the understanding that we explicitly want the provider's value and "unit" is the most logical name.

Reviewer 3: I think this is the pattern we should follow for all columns that are governed by an attribute TBH. That way we have a very clear way of saying in the content constraints, if the value format is X, this column must meet requirements specified in X for al it's values.

## Attribute ID

UnitFormat

## Attribute Name

Unit Format

## Description

> Indicates standards for expressing units of measure in columns appearing in billing data.

Reviewer 2: nit
```suggestion
Indicates standards for expressing measurement units in columns appearing in billing data.
```

* Unit values and components of columns using the Unit Format MUST use a capitalization scheme that is consistent with the capitalization scheme used in this attribute, if that term is listed in this section. For example, a value of "gigabyte-seconds" would not be compliant with this specification as the terms "gigabyte" and "second" are listed in this section with the appropriate capitalization.  If the unit is not listed in the table, it is to be used over a functional equivalent with similar meaning with the same capitalization scheme.
* Units SHOULD be composed of the list of recommended units listed in this section, unless the unit value covers a dimension not listed in the recommended unit set, or if the unit covers a count-based unit distinct from recommended values in the count dimension listed in this section.  

### Data Size Unit Names

Data size MUST be abbreviated using the following standard abbreviations.  Each data size abbreviation can be considered both the singular and plural form of the unit.  For example, "GB" is both the singular and plural form of the unit "GB".  The following table lists the valid data size units.

> | Data size in bits    | Data size in bytes    |

Reviewer: Is this an _abbreviation_ table, or a _valid data size_ table?  The previous paragraph implies both.  Can a provider list their data in bytes?

Reviewer 2: Based on previous conversations, do we want them to provide data in bytes? I want to say this came up in the practitioner squad that something that small isn't useful. For the record, I'm fine either way in 1.0.

| -------------------- | --------------------- |
| Kib (kibibit = 2^10) | KiB (kibibyte = 2^10) |
| Mib (mebibit = 2^20) | MiB (mebibyte = 2^20) |
| Gib (gibibit = 2^30) | GiB (gibibyte = 2^30) |
| Tib (tebibit = 2^40) | TiB (tebibyte = 2^40) |
| Pib (pebibit = 2^50) | PiB (pebibyte = 2^50) |
| Eib (exbibit = 2^60) | EiB (exbibyte = 2^60) |
| Kb (kilobit = 10^3)  | KB (kilobyte = 10^3)  |
| Mb (megabit = 10^6)  | MB (megabyte = 10^6)  |
| Gb (gigabit = 10^9)  | GB (gigabyte = 10^9)  |
| Tb (terabit = 10^12) | TB (terabyte = 10^12) |
| Pb (petabit = 10^15) | PB (petabyte = 10^15) |

> | Eb (exabit = 10^15)  | EB (exabyte = 10^15)  |

Reviewer: ```suggestion
| Eb (exabit = 10^18)  | EB (exabyte = 10^18)  |
```

### Count-based Unit Names

A count-based unit is a noun that represents a discrete number of items, events, or services.  For example, a count-based unit can be used to represent the number of requests, instances, tokens, or connections.  

> If the list of recommended values does not cover a count-based unit, a provider MAY introduce a new noun representing a count-based unit.  All nouns appearing in unit that are not listed in the recommended values table will be considered count-based units.  A new count-based unit value MUST be capitalized.

Reviewer: ```suggestion
If the following list of recommended values does not cover a count-based unit, a provider MAY introduce a new noun representing a count-based unit.  All nouns appearing in unit that are not listed in the recommended values table will be considered count-based units.  A new count-based unit value MUST be capitalized.
```

The judges would also accept "below" or something to link to the table.

| Count        |

> |--------------|

Reviewer: Is this an exhaustive list, or is it merely a list of examples?  Either way, it would be good to specify.

Reviewer: Resolving in favor of comment made on line 54.

| Count        |
| Unit         |
| Request      |
| Token        |
| Connection   |
| Certificate  |
| Domain       |
| Core         |

### Time-based Unit Names

A time-based unit is a noun that represents a time interval.  Time-based units can be used to measure consumption over a time interval or in combination with another unit to capture a rate of consumption.

| Time         |
|--------------|

> | Year         |

Reviewer: Same note as above: is this an exhaustive list, or is it merely a list of examples? Either way, it would be good to specify.

Reviewer 4: Ok, I added a sentence to clarify that units MUST match one of the values.

| Month        |
| Day          |
| Hour         |
| Minute       |
| Second       |

### Composite Units

If the unit value is a composite value made from combinations of one or more units, each component MUST also align with the set of recommended values.

> Instead of "per" or "-" to denote a Composite Unit, slash ("/") and space(" ") MUST be used as a common convention.  Count-based units like requests, instances, and tokens SHOULD be expressed using a value listed in the count dimension.  For example, if a usage unit is measured as a rate of requests or instances over a period of time, the unit SHOULD be listed as "count/day" to signify the number of requests per day.

Reviewer: "Requests" is in the list of recommended values.  Thus:

```suggestion
Instead of "per" or "-" to denote a Composite Unit, slash ("/") and space(" ") MUST be used as a common convention.  Count-based units like requests, instances, and tokens SHOULD be expressed using a value listed in the count dimension.  For example, if a usage unit is measured as a rate of requests or instances over a period of time, the unit SHOULD be listed as "request/day" to signify the number of requests per day.
```
